include/kernels/gauge_polyakov_loop.cuh

```diff
@@ -0,0 +1,221 @@
+    static constexpr QudaReconstructType recon = recon_;
+    using Gauge = typename gauge_mapper<Float,recon>::type;
+    using Link = Matrix<complex<Float>, 3>;
+
+    int X_bulk[4];
+    int_fastdiv X_slice[4];
+    Gauge U;
+    Gauge S;
+    int timeslice;
+
+    GaugeInsertTimesliceArg(GaugeField &U_, const GaugeField &S_, const int timeslice_) :
+      kernel_param(dim3(S_.LocalVolumeCB(), 2, 1)),
+      U(U_),
+      S(S_),
+      timeslice(timeslice_)
+    {
+      if (U_.Geometry() != QUDA_SCALAR_GEOMETRY || S_.Geometry() != QUDA_SCALAR_GEOMETRY)
+        errorQuda("Unexpected geometry pair U %d ; S %d", U_.Geometry(), S_.Geometry());
+      for (int dir = 0; dir < 4; dir++) {
+        X_bulk[dir] = U_.X()[dir];
+        X_slice[dir] = S_.X()[dir];
+        if (dir != 3 && U_.X()[dir] != S_.X()[dir]) errorQuda("Lengths %d %d in dimension %d do not agree", U_.X()[dir], S_.X()[dir], dir);
+      }
+      if (S_.X()[3] != 1) errorQuda("Unexpected time extent %d, expected 1", S_.X()[3]);
+    }
+
+  };
+
+  template <typename Arg> struct InsertTimeslice {
+    const Arg &arg;
+    constexpr InsertTimeslice(const Arg &arg) : arg(arg) {}
+    static constexpr const char *filename() { return KERNEL_FILE; }
+
+    __device__ __host__ inline void operator()(int x_cb, int parity)
+    {
+      int x[4];
+      getCoords(x, x_cb, arg.X_slice, parity);
+
+      // Get bulk x_cb/parity
+      x[3] = arg.timeslice;
+      int x_bulk_cb = linkIndex(x, arg.X_bulk);
+      int parity_bulk = (x[0] + x[1] + x[2] + x[3]) & 1;
+
+      // Load from slice, store into bulk
+      typename Arg::Link link = arg.S(0, x_cb, parity);
+      arg.U(0, x_bulk_cb, parity_bulk) = link;
+
+    }
+
+  };
+
+  /**
+     @brief Calculates the Polyakov loop in a given direction, returning the product matrix
+
+     @return The product of the gauge path
+     @param[in] arg Kernel argumnt
+     @param[in] x Full index array
+     @param[in] parity Parity index
+  */
+  template <int dir, typename Arg>
+  __device__ __host__ inline typename Arg::HighPrecLink
+  computePolyakovLoop(const Arg &arg, int x[4], int parity)
+  {
+    int dx[4] = {0, 0, 0, 0};
+
+    using Link = typename Arg::Link;
+    using HighPrecLink = typename Arg::HighPrecLink;
+
+    // polyloop: current matrix
+    // link: the loaded matrix in this round
+    Link link;
+    HighPrecLink hi_link, polyloop;
+    setIdentity(&polyloop);
+
+    int nbr_oddbit = parity;
+
+    for (int dt = 0; dt < arg.X[dir]; dt++) {
+      dx[dir] = dt;
+      link = arg.U(arg.geometry == QUDA_VECTOR_GEOMETRY ? dir : 0, linkIndexShift(x, dx, arg.X), nbr_oddbit);
+      hi_link = link; // promote
+      polyloop = polyloop * hi_link;
+      nbr_oddbit = nbr_oddbit ^ 1;
+    } // dt
+    return polyloop;
+  }
+
+  template <typename Float_, int nColor_, QudaReconstructType recon_>
+  struct GaugePolyakovLoopProductArg : public kernel_param<> {
+    using Float = Float_;
```

Comment:
As above: change this to using real = typename mapper<Float>::type;

Reply:
https://github.com/lattice/quda/commit/58845f78894fa8a26a831a6d0a61b8a50b56a269